core/src/main/scala/com/salesforce/op/evaluators/OpForecastEvaluator.scala

```diff
@@ -62,40 +67,105 @@ private[op] class OpForecastEvaluator
     log.info("Evaluated metrics: {}", metrics.toString)
     metrics
+
   }
 
-  protected def getSMAPE(data: Dataset[_], labelCol: String, predictionValueCol: String): Double = {
-    data.select(labelCol, predictionValueCol).rdd
-      .map(r => SMAPEValue(r.getAs[Double](0), r.getAs[Double](1)))
-      .reduce(_ + _).value
+
+  protected def getMetrics(data: Dataset[_], labelCol: String, predictionValueCol: String): ForecastMetrics = {
+    val res = data.select(labelCol, predictionValueCol).rdd
+      .map(r => MetricValue(r.getAs[Double](0), r.getAs[Double](1))).coalesce(1)
+      .fold(MetricValue.zero(seasonalWindow))(_ + _)
+    ForecastMetrics(
+      SMAPE = res.smape,
+      seasonalError = res.seasonalError,
+      MASE = res.mase
+    )
+
   }
 }
 
-/**
- * SMAPE value computation. See formula here:
- * https://www.m4.unic.ac.cy/wp-content/uploads/2018/03/M4-Competitors-Guide.pdf
- *
- * @param SMAPE symmetric Mean Absolute Percentage Error
- */
-object SMAPEValue {
-  def apply(y: Double, yHat: Double): SMAPEValue = {
-    SMAPEValue(2 * Math.abs(y - yHat), Math.abs(y) + Math.abs(yHat), 1L)
+// https://www.m4.unic.ac.cy/wp-content/uploads/2018/03/M4-Competitors-Guide.pdf
+object MetricValue {
+  def apply(y: Double, yHat: Double): MetricValue = {
+    val absDiff = Math.abs(y - yHat)
+    val sumAbs = Math.abs(y) + Math.abs(yHat)
+    val smapeSum = if (sumAbs > 0) {
+      absDiff / sumAbs
+    } else {
+      0.0
+    }
+    MetricValue(absDiff, smapeSum, SeasonalAbsDiff.apply(0, y), 1L)
+  }
+
+  def zero(seasonalWindow: Int): MetricValue = {
+    MetricValue(0d, 0d, SeasonalAbsDiff.zero(seasonalWindow), 0L)
   }
-  implicit val smapeSG: Semigroup[SMAPEValue] = caseclass.semigroup[SMAPEValue]
+
+  implicit val smapeSG: Semigroup[MetricValue] = caseclass.semigroup[MetricValue]
 }
 
-case class SMAPEValue private (nominator: Double, denominator: Double, cnt: Long) {
-  def value: Double = if (denominator == 0.0) Double.NaN else (nominator / denominator) / cnt
+case class MetricValue private(absDiff: Double, smapeSum: Double, seasonalAbsDiff: SeasonalAbsDiff, cnt: Long) {
+  def smape: Double = {
+    if (cnt == 0) {
+      Double.NaN
+    } else {
+      2 * smapeSum / cnt
+    }
+  }
+
+  def seasonalError: Double = seasonalAbsDiff.seasonalError(cnt)
+
+  def mase: Double = absDiff / (seasonalError * cnt)
+
+}
+
+object SeasonalAbsDiff {
+  def apply(seasonalWindow: Int, y: Double): SeasonalAbsDiff =
+    SeasonalAbsDiff(seasonalWindow, 1, mutable.ListBuffer[Double](y), 0.0)
+
+  def zero(seasonalWindow: Int): SeasonalAbsDiff =
+    SeasonalAbsDiff(seasonalWindow, 0, mutable.ListBuffer[Double](), 0.0)
+
+  implicit val seasonalAbsDiffSG: Semigroup[SeasonalAbsDiff] = new Semigroup[SeasonalAbsDiff] {
+    override def plus(x: SeasonalAbsDiff, y: SeasonalAbsDiff): SeasonalAbsDiff = x.combine(y)
+  }
 }
 
+case class SeasonalAbsDiff private
+(seasonalWindow: Int, lagsLength: Int, lags: mutable.ListBuffer[Double], absDiff: Double) {
```

Comment:
why do you need to keep `lagsLength`? you can do `lags.length` instead.

Reply:
I've just realized that ListBuffer::tail is O(N) and will have to replace that anyway.

Reply:
Perhaps use mutable.ArraryBuffer?